THis project is about used cars on eBay Kleinanzeigen, a classifieds section of the German eBay website, and the  aim of this project is to clean the dataset and perform some initial analysis on it.

In [1]:
import numpy as np
import pandas as pd
import re
autos = pd.read_csv("autos.csv", encoding="latin1")
# print(autos)

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

autos.info printed out all the info about the file  while the autos.heach() printed out the first 5 row of the file.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename({"dateCrawled": "date_crawled","offerType": "offer_type",
              "vehicleType": "vehicle_type","yearOfRegistration": "registration_year","monthOfRegistration":"registration_month","notRepairedDamage":"unrepaired_damage", "dateCreated": "ad_created", "postalCode": "postal_code", "nrOfPictures":"no_Pictures", "powerPS": "power_pS","lastSeen": "last_seen"
    
}, axis=1, inplace=True

) == autos.columns

autos.head()



date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_pS   model   odometer  registration_month fuelType  \
0    manuell       158  andere  150,000km                   3      lpg   
1  automatik       286     7er  150,000km                   6   benzin   
2    manuell       102    golf   70,000km                   7   benzin   
3  automatik        71  fortwo   70,000km                   6   benzin   
4    manuell         0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  no_Pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

After changing the column names from camelcase to snakecase using the .rename method on the column axis, the columns became readable

In [7]:
autos["odometer"].describe()

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object

In [8]:
autos[["price", "odometer"]].drop_duplicates()

price   odometer
0       $5,000  150,000km
1       $8,500  150,000km
2       $8,990   70,000km
3       $4,350   70,000km
4       $1,350  150,000km
...        ...        ...
49939  $16,995   80,000km
49941  $62,000   10,000km
49950   $3,099   30,000km
49990  $17,500   30,000km
49997  $13,200    5,000km

[7767 rows x 2 columns]

In [9]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(float)
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(float)
autos.rename({"odometer": "odometer_km"},axis=1, inplace=True)
autos[["price","odometer_km"]].drop_duplicates()




price  odometer_km
0       5000.0     150000.0
1       8500.0     150000.0
2       8990.0      70000.0
3       4350.0      70000.0
4       1350.0     150000.0
...        ...          ...
49939  16995.0      80000.0
49941  62000.0      10000.0
49950   3099.0      30000.0
49990  17500.0      30000.0
49997  13200.0       5000.0

[7767 rows x 2 columns]

In [10]:
autos[autos["price"].between(5000,  7000)]
autos[autos["odometer_km"].between(10000, 50000)]

date_crawled                                               name  \
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
42     2016-03-22 21:37:03  Vw_Polo_l.0_/60_PS_Blue_Motion_Technologie_Son...   
53     2016-04-02 17:45:01      Dacia_Sandero_Stepway_II_MPI_85__top_gepflegt   
...                    ...                                                ...   
49941  2016-03-15 00:53:25                   Maserati_Ghibli_Diesel_Automatik   
49950  2016-03-15 13:45:36         Volvo_V70_2_II__2002__AHK__Klima__Volleder   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...   
49990  2016-03-21 19:54:19         Mercedes_Benz_A_200__BlueEFFICIENCY__Urban   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   

       seller offer_type    price   abtest vehicle_type  registration_year  \
12     privat    Angebot   5299.0  control   kleinwagen               2010   
22     privat    Angebot  25450.0  control       cabrio               2015   
24     privat    Angebot  48500.0  control    limousine               2014   
42     privat    Angebot  11900.0  control   kleinwagen               2014   
53     privat    Angebot   9000.0  control    limousine               2012   
...       ...        ...      ...      ...          ...                ...   
49941  privat    Angebot  62000.0     test    limousine               2015   
49950  privat    Angebot   3099.0  control        kombi               2002   
49987  privat    Angebot  21990.0  control    limousine               2013   
49990  privat    Angebot  17500.0     test    limousine               2012   
49998  privat    Angebot  22900.0  control        kombi               2013   

         gearbox  power_pS     model  odometer_km  registration_month  \
12     automatik        71    fortwo      50000.0                   9   
22       manuell       184    cooper      10000.0                   1   
24     automatik       306       5er      30000.0                  12   
42       manuell        60      polo      20000.0                   7   
53       manuell        84   sandero      30000.0                   7   
...          ...       ...       ...          ...                 ...   
49941  automatik       275       NaN      10000.0                   8   
49950    manuell       140       NaN      30000.0                   5   
49987    manuell       150        a3      50000.0                  11   
49990    manuell       156  a_klasse      30000.0                  12   
49998    manuell       150        a3      40000.0                  11   

      fuelType           brand unrepaired_damage           ad_created  \
12      benzin           smart              nein  2016-03-31 00:00:00   
22      benzin            mini              nein  2016-03-28 00:00:00   
24      benzin             bmw              nein  2016-04-03 00:00:00   
42      benzin      volkswagen               NaN  2016-03-22 00:00:00   
53      benzin           dacia              nein  2016-04-02 00:00:00   
...        ...             ...               ...                  ...   
49941   diesel  sonstige_autos              nein  2016-03-14 00:00:00   
49950      NaN           volvo              nein  2016-03-15 00:00:00   
49987   diesel            audi              nein  2016-03-22 00:00:00   
49990   benzin   mercedes_benz              nein  2016-03-21 00:00:00   
49998   diesel            audi              nein  2016-03-08 00:00:00   

       no_Pictures  postal_code            last_seen  
12               0        34590  2016-04-06 14:17:52  
22               0        44789  2016-04-01 06:45:30  
24               0        22547  2016-04-07 13:16:50  
42               0        26629  2016-04-06 18:44:39  
53               0        53639  2016-04-06 16:46:50  
...            ...  

In [11]:
autos["date_crawled"].value_counts(normalize=True, dropna=False)


date_crawled
2016-03-19 17:36:18    0.00006
2016-03-30 19:48:02    0.00006
2016-03-21 16:37:21    0.00006
2016-03-27 22:55:05    0.00006
2016-03-21 20:37:19    0.00006
                        ...   
2016-03-28 07:36:23    0.00002
2016-03-31 22:49:18    0.00002
2016-04-02 19:50:18    0.00002
2016-03-15 18:57:06    0.00002
2016-04-01 12:45:07    0.00002
Name: proportion, Length: 48213, dtype: float64

In [12]:
autos["ad_created"].value_counts(normalize=True, dropna=False)


ad_created
2016-04-03 00:00:00    0.03892
2016-03-20 00:00:00    0.03786
2016-03-21 00:00:00    0.03772
2016-04-04 00:00:00    0.03688
2016-03-12 00:00:00    0.03662
                        ...   
2016-01-16 00:00:00    0.00002
2016-02-07 00:00:00    0.00002
2015-12-30 00:00:00    0.00002
2016-01-07 00:00:00    0.00002
2016-02-16 00:00:00    0.00002
Name: proportion, Length: 76, dtype: float64

In [13]:
autos["last_seen"].value_counts(normalize=True, dropna=False)


last_seen
2016-04-07 06:17:27    0.00016
2016-04-06 06:17:24    0.00014
2016-04-07 03:16:17    0.00014
2016-04-06 21:17:51    0.00014
2016-04-07 06:46:12    0.00012
                        ...   
2016-03-29 20:41:51    0.00002
2016-03-25 23:18:15    0.00002
2016-04-05 13:16:35    0.00002
2016-03-17 09:45:12    0.00002
2016-04-02 19:49:19    0.00002
Name: proportion, Length: 39481, dtype: float64

In [14]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Observations on date columns
All three date columns (`date_crawled`, `ad_created`, `last_seen`) show that the ads were created, crawled, and last seen between March and April 2016. Most activity is concentrated in this short window, with very few listings outside these months.


Using the .describe() the lowest is 1000 and the highest is 9999

In [15]:
auto = autos[autos["registration_year"].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

registration_year
2000    0.06708
2005    0.06030
1999    0.06000
2004    0.05474
2003    0.05454
2006    0.05416
2001    0.05406
2002    0.05066
1998    0.04906
2007    0.04608
Name: proportion, dtype: float64

In [16]:

#Exploring Price by Brand 
brands_counts = autos["brand"].value_counts(normalize=True)
common_brands = brands_counts[brands_counts > .05].index

In [17]:
brands_mean_prices = { }


for car_brands in common_brands:
    selected_brand = autos[autos["brand"] == car_brands]
    mean_price = selected_brand["price"].mean()
    brands_mean_prices[car_brands] = int(mean_price)
brands_mean_prices

{'volkswagen': 6384,
 'opel': 5106,
 'bmw': 8252,
 'mercedes_benz': 29511,
 'audi': 8965,
 'ford': 7105}

In [18]:
bmp_series = pd.Series(brands_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           6384
opel                 5106
bmw                  8252
mercedes_benz       29511
audi                 8965
ford                 7105

In [19]:
car_mean_mileage = { }
for car_brands in common_brands:
    selected_brand = autos[autos["brand"] == car_brands]
    mean_mileage = selected_brand["odometer_km"].mean()
    car_mean_mileage[car_brands] = int(mean_mileage)
car_mean_mileage

mean_mileage = pd.Series(car_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brands_mean_prices).sort_values(ascending=False)

In [20]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132521
mercedes_benz        130886
audi                 129643
opel                 129298
volkswagen           128955
ford                 124131

In [21]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132521        8252
mercedes_benz        130886       29511
audi                 129643        8965
opel                 129298        5106
volkswagen           128955        6384
ford                 124131        7105

In [22]:
print(autos[["fuelType","unrepaired_damage"]])

      fuelType unrepaired_damage
0          lpg              nein
1       benzin              nein
2       benzin              nein
3       benzin              nein
4       benzin              nein
...        ...               ...
49995   diesel              nein
49996   benzin              nein
49997   benzin              nein
49998   diesel              nein
49999   benzin              nein

[50000 rows x 2 columns]


In [23]:
# translate german to English

autos = autos.replace({
    "unrepaired_damage": {"nein": "No", "ja":"yes"},
    "fuelType": {"benzin": "diesel","diesel": "diesel"},
    "gearbox": {"manuell": "manual", "automatik": "automatic"},
    "vehicle_type": {"kombi": "wagon", "limousine": "sedan"}
    
    
    
    

})

print(autos)


# ang_correction1 = {
#     "nein": "No"
# }
# lang_correct2 = {
#     "benzin": "diesel",
#     "diesel": "diesel"
    
# }

# damage = autos["unrepaired_damage"].map(lang_correction1)

# fuel = autos["fuelType"].map(lang_correct2)
# print(damage, fuel)

              date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type   

In [24]:

# Convert the dates to be uniform numeric data
date_cols = ["date_crawled", "ad_created", "last_seen"]

for cols in date_cols:
    autos[cols] = autos[cols].astype(str).str.replace(r"\D", "", regex=True).astype("int64")
    
    print(autos)

         date_crawled                                               name  \
0      20160326174746                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      20160404133856         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      20160326185724                         Volkswagen_Golf_1.6_United   
3      20160312165810  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      20160401143850  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...               ...                                                ...   
49995  20160327143819   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  20160328105025  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  20160402144448                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  20160308192542                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  20160314004212                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  registration_year  \
0      pr

In [25]:
print(autos["name"])

0                         Peugeot_807_160_NAVTECH_ON_BOARD
1               BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik
2                               Volkswagen_Golf_1.6_United
3        Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...
4        Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...
                               ...                        
49995     Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon
49996    Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...
49997                       Fiat_500_C_1.2_Dualogic_Lounge
49998                   Audi_A3_2.0_TDI_Sportback_Ambition
49999                                  Opel_Vectra_1.6_16V
Name: name, Length: 50000, dtype: object


In [26]:
# extract the brands name
autos["new_name"] = autos["name"].str.split("_").str[0]
autos["new_name"] = autos["new_name"].str.lower()


# Check the result
print(autos["new_name"].head(20))



0        peugeot
1            bmw
2     volkswagen
3          smart
4           ford
5       chrysler
6             vw
7           golf
8           seat
9        renault
10            vw
11      mercedes
12         smart
13          audi
14       renault
15      corvette
16          opel
17    volkswagen
18     verkaufen
19         mazda
Name: new_name, dtype: object


In [27]:
# Find the most common brand/model combinations

# for brands
autos["brandss"] = autos["name"].str.split("_").str[0].str.lower()
# for models
autos["models"] = autos["name"].str.split("_").str[1].str.lower()
# combine model and brand
autos["brands_model"] = autos["brandss"] + "_" + autos["models"]

autos["brands_model"].value_counts().head(10)



brands_model
mercedes_benz        3261
volkswagen_golf      1852
opel_corsa           1447
opel_astra           1250
audi_a4              1113
audi_a6               754
audi_a3               751
volkswagen_passat     741
ford_focus            734
vw_golf               721
Name: count, dtype: int64

In [28]:
# 1. CREATE MILEAGE GROUPS
bins = [0, 50000, 100000, 150000, 200000, 250000, 300000, 350000, 400000, 500000]

labels = ["0-50k", "50k - 100k", "100k-150k", "150k-200k", "200k-250k", "250k-300k","300k-350k", "350k-400k", "400k-500k"]

# create a new column with mileage groups

autos["odometerkm_grps"] = pd.cut(autos["odometer_km"], bins=bins, labels=labels)


# 2. AGGREAGATE AVERAGE PRICE BY GROUP
#  Group by mileage group and calculate price by group

avg_price_by_mileage = autos.groupby("odometerkm_grps", observed=True)["price"].mean()

# sort for readability

avg_price_by_mileage = avg_price_by_mileage.sort_index()
print(avg_price_by_mileage)

odometerkm_grps
0-50k         23445.861720
50k - 100k    10629.222022
100k-150k      7994.325823
Name: price, dtype: float64


In [29]:
avg_price_by_damage = autos.groupby("unrepaired_damage")["price"].mean()
print(avg_price_by_damage)


unrepaired_damage
No     8470.595794
yes    2311.889249
Name: price, dtype: float64


In [30]:
price_no_damage = avg_price_by_damage["No"]
price_with_damage = avg_price_by_damage["yes"]

difference = price_no_damage - price_with_damage
percentage_diff = (difference / price_no_damage) * 100

print(f"Damaged cars are on average {difference:.2f} cheaper ({percentage_diff:.1f}% less) than non-damaged cars.")


Damaged cars are on average 6158.71 cheaper (72.7% less) than non-damaged cars.
